core/utils.py

```diff
@@ -58,3 +57,220 @@ def query_result_to_array(query_result, date_iso=True):
+
+
+def parse_date_range_argument(request_args):
+    """
+    Parses date range arguments from the request_arguments string.
+
+    Arguments:
+        request_args - request arguments as a string
+        arg - argument to be extracted from the request arguments
+
+    Returns:
+        tuple of two datetime objects
+    """
+
+    if request_args is None:
+        return get_default_datetime_range()
+
+    try:
+        dt_to = (
+            datetime.strptime(request_args.split("-")[1], "%Y%m%d").replace(
+                hour=0, minute=0, second=0, microsecond=0
+            )
+            + timedelta(days=1)
+            + timedelta(milliseconds=-1)
+        )
+
+        dt_from = datetime.strptime(request_args.split("-")[0], "%Y%m%d").replace(
+            hour=0, minute=0, second=0, microsecond=0
+        )
+
+        return dt_from, dt_to
+
+    except ValueError:
+        return get_default_datetime_range()
+
+
+def download_csv(readings, filename_base="results"):
+    """
+    Use Pandas to convert array of readings into a csv
+    Args:
+       readings: a list of records to be written out as csv
+       filename (optional): str, name of downloaded file
+    Returns:
+        send_file: function call to flask send_file, will send csv file to client.
+    """
+    df = pd.DataFrame(readings)
+    output_buffer = io.BytesIO()
+    df.to_csv(output_buffer)
+    output_buffer.seek(0)
+    filename = (
+        filename_base + "_" + datetime.now().strftime("%d-%m-%Y_%H-%M-%S") + ".csv"
+    )
+    return send_file(
+        output_buffer, download_name=filename, mimetype="text/csv", as_attachment=True
+    )
+
+
+def filter_latest_sensor_location(db):
+    """Return a filter object that excludes all but the latest location for each sensor.
+
+    This should be used to filter a query that involves the SensorLocationClass.
+
+    Args:
+        db: A database connection.
+    Returns:
+        An object that can be given as an argument to sqlalchemy.filter.
+    """
+    query = (
+        db.session.query(
+            SensorLocationClass.sensor_id,
+            func.max(SensorLocationClass.installation_date).label("installation_date"),
+        )
+        .group_by(SensorLocationClass.sensor_id)
+        .subquery()
+    )
+    return and_(
+        query.c.sensor_id == SensorLocationClass.sensor_id,
+        query.c.installation_date == SensorLocationClass.installation_date,
+    )
+
+
+def vapour_pressure_deficit(temperature, relative_humidity):
+    """Compute vapour pressure deficit from T&RH data.
+
+    Args:
+        temperature: Temperature in celsius. Must support element-wise np.exp, so can be
+            a scalar or e.g. a numpy array.
+        relative_humidity: Relative humidity in percentage. Can be a scalar or an array
+            of the same length as temperature.
+
+    Returns:
+    Vapour pressure deficit, in pascals.
+    """
+    # See https://pulsegrow.com/blogs/learn/vpd for the magic formula.
```

Comment:
I would also add this reference: https://en.wikipedia.org/wiki/Tetens_equation
And state that this is Tetens Equation

Reply:
Done